src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -65231,6 +65231,60 @@ void M(string? s)
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(9, 13));
         }
 
+        [Fact]
+        [WorkItem(38638, "https://github.com/dotnet/roslyn/issues/38638")]
+        public void ForEach_24()
+        {
+            var source =
+@"#nullable enable
+using System.Diagnostics.CodeAnalysis;
+struct S<T>
+{
+    public E<T> GetEnumerator() => new E<T>();
+}
+struct E<T>
+{
+    [MaybeNull]public T Current => default;
+    public bool MoveNext() => false;
+}
+class Program
+{
+    static T F1<T>()
+    {
+        foreach (var t1 in new S<T>())
+            return t1; // 1
+        foreach (T t2 in new S<T>())
+            return t2; // 2
+        throw null!;
+    }
+    static object F2<T>()
+    {
+        foreach (object o1 in new S<T>()) // 3
+            return o1; // 4
+        foreach (object? o2 in new S<T>())
+            return o2; // 5
+        throw null!;
+    }
+}";
+            var comp = CreateCompilation(new[] { source, MaybeNullAttributeDefinition });
+            comp.VerifyDiagnostics(
+                // (17,20): warning CS8603: Possible null reference return.
+                //             return t1; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "t1").WithLocation(17, 20),
+                // (19,20): warning CS8603: Possible null reference return.
+                //             return t2; // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "t2").WithLocation(19, 20),
+                // (24,25): warning CS8606: Possible null reference assignment to iteration variable
```

Comment:
>CS8606 [](start = 36, length = 6)

probably should not have this warning per our previous discussion (and this should be a W warning, but we decided to suppress in the case of [MaybeNull]T state)

Reply:
There should be a warning here because the target type can be specified explicitly (as `object?`). But it should be a W-warning as you suggested. Logged [https://github.com/dotnet/roslyn/issues/40164](https://github.com/dotnet/roslyn/issues/40164).

---
In reply to: [354458819](https://github.com/dotnet/roslyn/pull/39983#discussion_r354458819) [](ancestors = 354458819)

Reply:
Oh, I see. We only suppress that assignment warning when the target is a `T` (but not if it's `object`). Thanks

---
In reply to: [354482412](https://github.com/dotnet/roslyn/pull/39983#discussion_r354482412) [](ancestors = 354482412,354458819)